_posts/2017-05-30-database-encryption.markdown

```diff
@@ -0,0 +1,70 @@
+---
+layout: post
+title: "Database encryption"
+date: 2017-05-30 11:03
+comments: true
+author: ashkan
+categories: [Databse, Encryption, Security]
+---
+
+Recently after examining the data stored in one of our systems, we noticed that while originally this system wasn't designed to include sensative data, over the time it ended up including some sensative information and we need to encrypt the data.
+
+While our systems were safe by other means, because of the issue above we decided to encrypt the database behind it to cover the cases when people end up accessing our database directly.
+
+# Our goal
+Encrypt database without any downtime.
+
+<!-- more -->
+
+# Approach
+In some ways this was a moving target, while we want to encrypt existing rows, new rows are constantly being added to our database. Because of that we ended up coming up with following steps:
+
+1. Add new _encrypted_ fields to the database
+2. Start populating new _encrypted_ fields with _encrypted_ values while still populating _non-encrypted_ fields
+3. migrate old rows by populating their _encrypted_ fields
+4. switch to use _encrypted_ fields
+5. drop _non-encrypted_ fields
+
+Each step above is relatively simple and can be tested properly before moving to the next step.
+
+## Encryption Libraries and decision
+We looked at few gems, mainly [attr_encrypted](https://github.com/attr-encrypted/attr_encrypted), [crypt_keeper](https://github.com/jmazzi/crypt_keeper) and [symmetric_encryption](https://github.com/rocketjob/symmetric-encryption). We looked at their documentation and also their stats in [RubyToolbox](https://www.ruby-toolbox.com/categories/encryption).
+
+In the end, we decided to use **Symmetric Encryption**, mainly based on their robust [docs](https://rocketjob.github.io/symmetric-encryption/), ease of use and easy integration with other libraries (in our case ActiveRecord). They provide some useful [rake tasks](https://rocketjob.github.io/symmetric-encryption/rake_tasks.html) and also describe how to configure this gem for apps inside/outside Heroku. Overall `symmetric_encryption` seemed really robust, reliable and 🔒 .
```

Comment:
Where did you base these last few things on, just docs or did you review the code?